docs/proposals/enable-local-source-code-support.md

```diff
@@ -0,0 +1,302 @@
+
+This EP is merelely concentrated on [1], but will make [2] easily to implement at a later stage.
+
+The API changes should consider the following features:
+
+- **Ability to upload/download source code**: Users should be able to specify how to build a bundle image. Users should be able to specify which bundle image to use, if it already exists.
+
+- **Ability to provide means of authentication to a registry**: Users should be able to authenticate when pulling/pushing to their container registry of choice.
+
+- **Ability to clean-up images if desired**: Users should have means to delete bundle images after the application container image is build.
+
+- **Ability to specify what to bundle**: Users should be able to specify an absolute path to their local source code directory.
+
+- **Ability to ensure reproducible Builds for bundle images**: Users should be able to verify if source code corresponds to a particular bundle image. To achieve this we can apply the pattern of reproducible builds, by pinning images timestamps to a particular time in the past. An example of the concept on reproducible builds is explained in [here](https://reproducible-builds.org/).
+
+- **Ability to specify particular directories that should not be bundle**: Users should be able to signalize which directories to not bundle, e.g. _/vendor_ . Reasonable defaults like _.github_ should be in place. Similar to the definitions under a `.gitignore` or `.dockerignore` file.
+
+### Proposal 01: API modifications
+
+> Extend `spec.source`
+
+ Introduce `spec.source.container`, which should host all the necessary metadata to signalize the usage of local source code, this should be a go `struct` which indicates what to bundle.
+
+ See an API [example](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L46-L59) . The [Container](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L46) struct should **at least** allow users to define the following:
+
+ 1. A `spec.source.container.credentials` optional [field](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L54) for defining a Kubernetes secret, in order to authenticate to the registry when pulling a bundle image. If not specified, it should use the same credentials as in `spec.output.credentials`.
+
+ 2. A `spec.source.container.path` mandatory [field](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L58) for specifying where the source code is located in the local workstation.
+
+ 3. A `spec.source.container.image` mandatory [field](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L49) for specifying the container image endpoint, registry and repository to build. If an image is pinned to a _SHA_, we could assume the image already exists, and we will only _pull_ it, and not longer _push_ it.
```

Comment:
I'd like to recommend that we only support images references by digest here, since references by tags can introduce race conditions. If we support referencing by tags, we should have some way to report the digest of that image in `.status`, so just only accepting refs by digest simplifies that.

Reply:
yes, thanks for highlight that. I agree. Done.